### PR TITLE
posix: fix unix socket accept/connect, add their nonblocking variants, handle MSG_PEEK in recv

### PR DESCRIPTION
**Pull request by adamgreloch**
JIRA: RTOS-877

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [X] New test added: (add PR link here).
https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/364
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: posix/unix.c [hunk 1]
```diff
@@ -373,7 +378,12 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
 			break;
 		}
 
-		if ((new = unixsock_alloc(&newid, s->type, nonblock)) == NULL) {
+		if (s->nonblock && (s->connect == NULL)) {
+			err = -EWOULDBLOCK;
+			break;
+		}
+
+		if ((new = unixsock_alloc(&newid, s->type, s->nonblock)) == NULL) {
```
Comment (agkaminski) [inline posix/unix.c:386]: I think it would be best to create new unixsock with `refcnt = 2` and require `unixsock_put()` at the end.
Rationale - we have a reference to the new object, that might be removed from the tree by someone else meanwhile. If we handle and use this reference, it should be reflected in the refcnt (1 for presence in the tree + 1 for handling it by us after creation).

File: posix/unix.c [hunk 2]
```diff
@@ -567,13 +580,23 @@ int unix_connect(unsigned socket, const struct sockaddr *address, socklen_t addr
...
+			}
+
+			while (s->connect == NULL && !(s->state & US_PEER_CLOSED))
 				proc_threadWait(&s->queue, &s->spinlock, 0, &sc);
 
 			s->state &= ~US_CONNECTING;
-			hal_spinlockClear(&s->spinlock, &sc);
 
-			err = EOK;
+			if (s->state & US_PEER_CLOSED)
```
Comment (agkaminski) [inline posix/unix.c:594]: Pls take a look at https://docs.phoenix-rtos.com/coding/index.html

File: posix/unix.c (hunk 1 above)
Comment (adamgreloch) [reply posix/unix.c]: Fair point. I've extended `unixsock_alloc` with an additional `refs` arg to pass `refs = 2` here.

File: posix/unix.c (hunk 2 above)
Comment (adamgreloch) [reply posix/unix.c]: Oops, sorry. Corrected.

File: posix/unix.c [hunk 3]
```diff
@@ -674,6 +674,11 @@ static ssize_t recv(unsigned socket, void *buf, size_t len, int flags, struct so
 		for (;;) {
 			proc_lockSet(&s->lock);
 			if (s->type == SOCK_STREAM) {
+				if (flags & MSG_PEEK) {
```
Comment (ziemleszcz) [inline posix/unix.c:677]: MSG_PEEK applies to packet sockets as well. Please add TODO if you don't want to implement this case.

File: posix/unix.c [hunk 4]
```diff
@@ -674,6 +674,11 @@ static ssize_t recv(unsigned socket, void *buf, size_t len, int flags, struct so
...
+					err = _cbuffer_peek(&s->buffer, buf, len);
+					proc_lockClear(&s->lock);
+					break;
```
Comment (ziemleszcz) [inline posix/unix.c:680]: Don't not break here - control data (`fdpass_unpack`) should be read too. Not sure if peek mode applies to control data, you should probably check how it works on Linux (see unix socket tests for sample code which send control data).

File: posix/unix.c [hunk 5]
```diff
@@ -567,6 +582,12 @@ int unix_connect(unsigned socket, const struct sockaddr *address, socklen_t addr
 			hal_spinlockSet(&s->spinlock, &sc);
 			s->state |= US_CONNECTING;
 
+			if (s->nonblock && (s->connect == NULL)) {
+				hal_spinlockClear(&s->spinlock, &sc);
+				err = -EINPROGRESS;
```
Comment (ziemleszcz) [inline posix/unix.c:587]: Please note that in this case we should handle `SO_ERROR` in `unix_getsockopt()` to allow user to detrmine whether `connect()` succeeded or failed (see man page on `connec()t` for details).

File: posix/unix.c [hunk 6]
```diff
@@ -997,21 +1018,24 @@ int unix_poll(unsigned socket, short events)
 	else {
 		if (events & (POLLIN | POLLRDNORM | POLLRDBAND)) {
 			proc_lockSet(&s->lock);
-			if (_cbuffer_avail(&s->buffer) > 0)
+			if (_cbuffer_avail(&s->buffer) > 0 || (s->nonblock && (s->connect != NULL))) {
```
Comment (ziemleszcz) [inline posix/unix.c:1026]: Just for clarity could you explain why this change is needed?

File: posix/unix.c [hunk 7]
```diff
@@ -997,21 +1018,24 @@ int unix_poll(unsigned socket, short events)
...
+			}
 			proc_lockClear(&s->lock);
 		}
 
 		if (events & (POLLOUT | POLLRDNORM | POLLRDBAND)) {
 			if ((conn = unixsock_get_connected(s)) != NULL) {
 				proc_lockSet(&conn->lock);
 				if (conn->type == SOCK_STREAM) {
-					if (_cbuffer_free(&conn->buffer) > 0)
+					if (_cbuffer_free(&conn->buffer) > 0) {
```
Comment (ziemleszcz) [inline posix/unix.c:1036]: You should probably add similar change as in above POLLIN case. Poll may be used to wait for non-blocking `connect()` completion (see man page on `connect()` for details).

Comment (adamgreloch) [reply posix/unix.c:1036]: Non-blocking `connect()` completion is handled already by `unixsock_get_connected()` few lines above, so here just checking whether `conn` bufer is ready for writing suffices

File: posix/unix.c (hunk 6 above)
Comment (adamgreloch) [reply posix/unix.c:1026]: Otherwise it would not be possible to detect pending connections that could be `accept()`-ed, *but* this is unclear and actually does not work if accept is set to blocking but one decides to poll the connections before calling it. Pushing a better change in a second

File: posix/unix.c (hunk 5 above)
Comment (adamgreloch) [reply posix/unix.c:593]: Done

File: posix/unix.c [hunk 8]
```diff
@@ -373,7 +377,12 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
 			break;
 		}
 
-		if ((new = unixsock_alloc(&newid, s->type, nonblock)) == NULL) {
+		if (s->nonblock && (s->connect == NULL)) {
+			err = -EWOULDBLOCK;
+			break;
+		}
+
+		if ((new = unixsock_alloc(&newid, s->type, s->nonblock, 2)) == NULL) {
```
[the remark this answers is not in the document]
Comment (adamgreloch) [reply posix/unix.c]: Oh, right. Done

File: posix/unix.c (hunk 3 above)
Comment (adamgreloch) [reply posix/unix.c:692]: Added a TODO

File: posix/unix.c (hunk 4 above)
Comment (adamgreloch) [reply posix/unix.c:695]: It looks like under linux control data is peeked (not read) as well:
https://github.com/torvalds/linux/blob/master/net/unix/af_unix.c#L2620

Following our discussion, I've added a TODO for this

File: posix/unix.c [hunk 9]
```diff
@@ -373,7 +380,12 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
 			break;
 		}
 
-		if ((new = unixsock_alloc(&newid, s->type, nonblock)) == NULL) {
+		if (s->nonblock && (s->connect == NULL)) {
+			err = -EWOULDBLOCK;
+			break;
+		}
+
+		if ((new = unixsock_alloc(&newid, s->type, nonblock, 2)) == NULL) {
```
Comment (ziemleszcz) [inline posix/unix.c:388]: Why are you setting ref counter to  2?

Comment (agkaminski) [reply posix/unix.c:388]: See my previous comment - there was a race condition on socket creation. We add the socket to the tree (ref++) and also return reference to it to the creator (ref++). So initial ref counter should be set 2 references

Comment (agkaminski) [reply posix/unix.c:388]: Perhaps a brief comment might clarify that for the future, though

Comment (ziemleszcz) [reply posix/unix.c:388]: What race condition? Do you mean this line:
https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/572/files#diff-1c9efdfe25b04087e6b3c413cc19b4236acb47caf506607b871bffdb51344898R422

File: posix/unix.c (hunk 4 above)
Comment (adamgreloch) [reply posix/unix.c:693]: It was above function declaration, possibly got hidden in the diff. Moved it inside `if`

File: posix/unix.c [hunk 10]
```diff
@@ -620,6 +647,14 @@ int unix_getsockopt(unsigned socket, int level, int optname, void *optval, sockl
 					err = -EINVAL;
 				}
 				break;
+			case SO_ERROR:
+				if (s->connect != NULL) {
+					err = 0;
```
[the remark this answers is not in the document]
Comment (adamgreloch) [reply posix/unix.c]: Corrected and added a comment. Yes, `SO_ERROR` should clear errors, but this applies to its general purpose outside of non-blocking connect check which I didn't intend to implement here. Added a TODO

File: posix/unix.c [hunk 11]
```diff
@@ -393,16 +405,18 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
 			proc_threadWait(&s->queue, &s->spinlock, 0, &sc);
 
 		LIST_REMOVE(&s->connect, conn);
-		s->state &= ~US_ACCEPTING;
+		s->state &= ~(US_ACCEPTING | US_CAN_ACCEPT);
```
[the remark this answers is not in the document]
Comment (adamgreloch) [reply posix/unix.c]: If connect was done before accept, it would have set the US_CAN_ACCEPT flag on remote. I was only unsetting this flag here. But yes, now that I've corrected the handling to use `s->connect`, it is not needed now

File: posix/unix.c (hunk 9 above)
Comment (agkaminski) [reply posix/unix.c:389]: > I think it would be best to create new unixsock with refcnt = 2 and require unixsock_put() at the end.
Rationale - we have a reference to the new object, that might be removed from the tree by someone else meanwhile. If we handle and use this reference, it should be reflected in the refcnt (1 for presence in the tree + 1 for handling it by us after creation).

File: posix/unix.c [hunk 12]
```diff
@@ -997,21 +1032,24 @@ int unix_poll(unsigned socket, short events)
 	else {
 		if (events & (POLLIN | POLLRDNORM | POLLRDBAND)) {
 			proc_lockSet(&s->lock);
-			if (_cbuffer_avail(&s->buffer) > 0)
+			if (_cbuffer_avail(&s->buffer) > 0 || (s->state & US_CAN_ACCEPT)) {
```
[the remark this answers is not in the document]
Comment (adamgreloch) [reply posix/unix.c]: Done. Regarding whether this should be for non-blocking sockets only, `accept` man doesn't imply that poll events should be delivered only in their case, so I think it's appropriate to deliver this event in both cases

File: posix/unix.c [hunk 13]
```diff
@@ -653,9 +687,16 @@ static ssize_t recv(unsigned socket, void *buf, size_t len, int flags, struct so
 		for (;;) {
 			proc_lockSet(&s->lock);
 			if (s->type == SOCK_STREAM) {
+				if (flags & MSG_PEEK) {
+					/* TODO: peek control data */
+					err = _cbuffer_peek(&s->buffer, buf, len);
+					proc_lockClear(&s->lock);
+					break;
```
Comment (ziemleszcz) [inline posix/unix.c:694]: You shouldn't break here. The code from lines 669-728 also applies to the peek `recv()`. Blocking peek `recv()` should be supported. Additionally, note that the non-blocking peek `recv()` currently returns 0 instead of `EAGAIN` when cbuffer is empty. Could you also add support for `MSG_DONTWAIT`? It should be fairly straightforward to implement once `recv()` is fixed.

Comment (adamgreloch) [reply posix/unix.c]: Corrected. If by `MSG_DONTWAIT` support you mean proper behavior when `MSG_DONTWAIT`, `MSG_PEEK` are both set, then it's also done

File: lib/cbuffer.c [hunk 14]
```diff
@@ -56,6 +56,17 @@ int _cbuffer_write(cbuffer_t *buf, const void *data, size_t sz)
 
 
 int _cbuffer_read(cbuffer_t *buf, void *data, size_t sz)
+{
+	int bytes = _cbuffer_peek(buf, data, sz);
+
+	buf->r = (buf->r + bytes) & (buf->sz - 1);
+	buf->full = 0;
```
[the remark this answers is not in the document]
Comment (adamgreloch) [reply lib/cbuffer.c]: Corrected

File: posix/unix.c [hunk 15]
```diff
@@ -394,15 +404,19 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
...
 		s->state &= ~US_ACCEPTING;
+		hal_spinlockClear(&s->spinlock, &sc);
 
 		/* FIXME: handle connecting socket removal */
 
-		conn->state &= ~US_PEER_CLOSED;
+		hal_spinlockSet(&conn->spinlock, &sc);
+
+		/* clear US_CONNECTING as well in case connect was non-blocking */
+		conn->state &= ~(US_PEER_CLOSED | US_CONNECTING);
```
Comment (ziemleszcz) [inline posix/unix.c:414]: Clearing US_CONNECTING in `unix_connect` is no longer necessary.

File: posix/unix.c [hunk 16]
```diff
@@ -223,9 +223,13 @@ static unixsock_t *unixsock_get_connected(unixsock_t *s)
 	unixsock_t *r;
 
 	proc_lockSet(&unix_common.lock);
+
 	r = s->connect;
-	if (r != NULL)
+
```
Comment (ziemleszcz) [inline posix/unix.c:228]: Unnecessary empty line.

File: posix/unix.c [hunk 17]
```diff
@@ -997,21 +1037,24 @@ int unix_poll(unsigned socket, short events)
 	else {
 		if (events & (POLLIN | POLLRDNORM | POLLRDBAND)) {
 			proc_lockSet(&s->lock);
-			if (_cbuffer_avail(&s->buffer) > 0)
+			if (_cbuffer_avail(&s->buffer) > 0 || (s->connect != NULL && (s->state & US_LISTENING))) {
```
Comment (ziemleszcz) [inline posix/unix.c:1040]: MISRA guideline: use `(s->state & US_LISTENING) != 0` style here and elsewhere. In general control expressions should have Boolean type.

File: lib/cbuffer.h [hunk 18]
```diff
@@ -56,4 +56,6 @@ extern int _cbuffer_read(cbuffer_t *buf, void *data, size_t sz);
 extern int _cbuffer_write(cbuffer_t *buf, const void *data, size_t sz);
 
 
-#endif
\ No newline at end of file
+extern int _cbuffer_peek(cbuffer_t *buf, void *data, size_t sz);
```
Comment (ziemleszcz) [inline lib/cbuffer.h:59]: You can use `const cbuffer_t *buf` since `peek` shouldn't modify the buffer.

File: posix/unix.c [hunk 19]
```diff
@@ -373,7 +372,14 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
 			break;
 		}
 
-		if ((new = unixsock_alloc(&newid, s->type, nonblock)) == NULL) {
+		if (s->nonblock && (s->connect == NULL)) {
```
Comment (ziemleszcz) [inline posix/unix.c:375]: MISRA guideline: `s->nonblock != 0` or `s->nonblock == 1`

File: posix/unix.c [hunk 20]
```diff
@@ -373,7 +372,14 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
...
+			err = -EWOULDBLOCK;
+			break;
+		}
+
+		/* alloc new socket with 2 refs: first ref for presence in the tree, second
+		 * ref returned to the caller */
+		if ((new = unixsock_alloc(&newid, s->type, nonblock, 2)) == NULL) {
```
Comment (ziemleszcz) [inline posix/unix.c:382]: MISRA guideline. Please use this style here and elsewhere:
x = f(...)
if (x == NULL) {
  ...
}

File: include/errno.h [hunk 21]
```diff
@@ -69,5 +69,7 @@
 #define EISCONN         106
 #define ENOTCONN        107
 #define ECONNREFUSED    111
+#define EALREADY        114
```
Comment (lukileczo) [inline include/errno.h:72]: Not doing full review, just a comment: as you're adding new definitions in kernel, please also remove the corresponding ones in `libphoenix` - https://github.com/phoenix-rtos/libphoenix/blob/0fbc038c6699bdc184ced7cdc979b194a2474f06/include/errno.h#L53

Comment (adamgreloch) [reply include/errno.h:72]: Thanks for mentioning this. Submitted a PR

File: posix/unix.c [hunk 22]
```diff
@@ -290,8 +292,10 @@ int unix_socket(int domain, int type, int protocol)
 	if (protocol != PF_UNSPEC)
 		return -EPROTONOSUPPORT;
 
-	if ((s = unixsock_alloc(&id, type, nonblock)) == NULL)
+	s = unixsock_alloc(&id, type, nonblock);
+	if (s == NULL)
```
Comment (ziemleszcz) [inline posix/unix.c:296]: MISRA: use curly braces.

File: posix/unix.c [hunk 23]
```diff
@@ -313,23 +317,30 @@ int unix_socketpair(int domain, int type, int protocol, int sv[2])
 	if (protocol != PF_UNSPEC)
 		return -EPROTONOSUPPORT;
 
-	if ((s[0] = unixsock_alloc(&id[0], type, nonblock)) == NULL)
+	s[0] = unixsock_alloc(&id[0], type, nonblock);
+	if (s[0] == NULL)
```
Comment (ziemleszcz) [inline posix/unix.c:321]: MISRA: use curly braces.

File: posix/unix.c [hunk 24]
```diff
@@ -539,6 +566,8 @@ int unix_connect(unsigned socket, const struct sockaddr *address, socklen_t addr
 			break;
 		}
 
+		// FIXME: caller may block indefinitely if remote gets closed after
```
Comment (ziemleszcz) [inline posix/unix.c:569]: Change to C style comment.

File: posix/unix.c [hunk 25]
```diff
@@ -653,24 +697,34 @@ static ssize_t recv(unsigned socket, void *buf, size_t len, int flags, struct so
 		for (;;) {
 			proc_lockSet(&s->lock);
 			if (s->type == SOCK_STREAM) {
-				err = _cbuffer_read(&s->buffer, buf, len);
+				if (peek) {
```
Comment (ziemleszcz) [inline posix/unix.c:700]: MISRA: `peek == 1`.

File: posix/unix.c [hunk 26]
```diff
@@ -653,24 +697,34 @@ static ssize_t recv(unsigned socket, void *buf, size_t len, int flags, struct so
...
 				_cbuffer_read(&s->buffer, &rlen, sizeof(rlen));
 				_cbuffer_read(&s->buffer, buf, err = min(len, rlen));
 
 				if (len < rlen)
 					_cbuffer_discard(&s->buffer, rlen - len);
 			}
-			if (err > 0 && control && controllen && *controllen > 0)
-				fdpass_unpack(&s->fdpacks, control, controllen);
+			/* TODO: peek control data */
+			if (!peek) {
```
Comment (ziemleszcz) [inline posix/unix.c:716]: MISRA: `peek != 1` or `peek == 0`

File: posix/unix.c [hunk 27]
```diff
@@ -653,24 +697,34 @@ static ssize_t recv(unsigned socket, void *buf, size_t len, int flags, struct so
...
+				if (err > 0 && control && controllen && *controllen > 0)
```
Comment (ziemleszcz) [inline posix/unix.c:717]: MISRA: use curly braces.

File: posix/unix.c [hunk 28]
```diff
@@ -653,24 +697,34 @@ static ssize_t recv(unsigned socket, void *buf, size_t len, int flags, struct so
...
+					fdpass_unpack(&s->fdpacks, control, controllen);
+			}
 			proc_lockClear(&s->lock);
 
 			if (err > 0) {
-				hal_spinlockSet(&s->spinlock, &sc);
-				proc_threadWakeup(&s->writeq);
-				hal_spinlockClear(&s->spinlock, &sc);
-
+				if (!peek) {
```
Comment (ziemleszcz) [inline posix/unix.c:723]: MISRA: `peek != 1` or` peek == 0`

File: posix/unix.c [hunk 29]
```diff
@@ -182,7 +182,9 @@ static unixsock_t *unixsock_alloc(unsigned *id, int type, int nonblock)
 	proc_lockInit(&r->lock, &proc_lockAttrDefault, "unix.socket");
 
 	r->id = *id;
-	r->refs = 1;
+	/* alloc new socket with 2 refs: first ref for presence in the tree, second
```
Comment (ziemleszcz) [inline posix/unix.c:185]: Maybe expand the comment to:
"one reference is for the socket's presence in the tree, and the second is for handling by the caller before returning the socket to the user (to protect against accidental removal by someone else)"

**Review by ziemleszcz — APPROVED** [review]
LGTM
